### PR TITLE
add gf.geometry.fillet and gf.geometry.boolean_polygons

### PR DESCRIPTION
**Pull request by joamatab**
- fixes https://github.com/gdsfactory/gdsfactory/issues/1428
- remove some deprecated code options that do nothing and remained since the gdstk to gdspy transition `num_divisions`

### CODE REVIEW

File: gdsfactory/geometry/fillet.py [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+    Args:
+        operand: polygon, Component, .
+        radius, (number or sequence) – Fillet radius.
+            A sequence of values can be used to define the fillet radius for each vertex.
+        tolerance: Tolerance used for calculating the polygonal approximation of the filleted corners.
+
+    """
+
+    if hasattr(operand, "get_polygons"):
+        operand = operand.get_polygons(as_array=False)
```
Comment (95-martin-orion) [inline gdsfactory/geometry/fillet.py:26]: For a `Component`, I would expect `fillet` to not only modify the geometry contained directly by that component, but also any geometry contained by subcomponents.

File: gdsfactory/geometry/invert.py [hunk 2]
```diff
@@ -4,15 +4,14 @@
...
-from gdsfactory.typings import Int2, LayerSpec, Union
+from gdsfactory.typings import LayerSpec
 
 
 @gf.cell
 def invert(
     elements,
     border: float = 10.0,
     precision: float = 1e-4,
-    num_divisions: Union[int, Int2] = (1, 1),
```
Comment (95-martin-orion) [inline gdsfactory/geometry/invert.py:15]: Is this change intentional? It's not listed in the PR description.

File: gdsfactory/geometry/union.py [hunk 3]
```diff
@@ -65,9 +64,13 @@ def union(
             U.add_polygon(unioned_polygons, layer=layer)
     else:
         all_polygons = component.get_polygons(by_spec=False)
-        unioned_polygons = _union_polygons(
-            all_polygons,
-            precision=precision,
+        unioned_polygons = (
```
Comment (95-martin-orion) [inline gdsfactory/geometry/union.py:67]: Ditto here - this isn't listed in the PR description.

If this is a separate cleanup, IMO it should be a separate PR.

**Review by 95-martin-orion — CHANGES_REQUESTED** [review]
A couple of comments, but the `fillet` method added here is enough to resolve #1428.

File: gdsfactory/geometry/invert.py (hunk 2 above)
Comment (joamatab) [reply gdsfactory/geometry/invert.py:15]: this is some deprecated code that does nothing and remained since the gdstk to gdspy transition

File: gdsfactory/geometry/union.py (hunk 3 above)
Comment (joamatab) [reply gdsfactory/geometry/union.py:67]: same here,

this is some deprecated code that does nothing and remained since the gdstk to gdspy transition

File: gdsfactory/geometry/fillet.py (hunk 1 above)
Comment (joamatab) [reply gdsfactory/geometry/fillet.py:26]: yes, get_polygons should give you the subcomponent polygons

**Comment by joamatab** [comment]
something a bit strange from this

```
    o = gf.components.mzi()
    p = fillet(o, radius=0.3)
    c = gf.Component("demo")
    c.add_polygon(p)
    c.show(show_ports=True)
```

it does not fillet  some polyons

image


@lucas-flexcompute

**Comment by joamatab** [comment]
I released v6.68.0, so we can play with it and see how it goes

**Comment by 95-martin-orion** [comment]
> something a bit strange from this {...}

The un-filleted section is a `Path`, and therefore not affected by `fillet`:
https://github.com/gdsfactory/gdsfactory/blob/6e38b04569c1ed747b77ea33f074e2fcbe06ad8c/gdsfactory/path.py#L1097-L1104

**Comment by heitzmann** [comment]
> something a bit strange from this
> 
> ```
>     o = gf.components.mzi()
>     p = gf.geometry.fillet(o, radius=0.3)
>     c = gf.Component("demo")
>     c.add_polygon(p)
>     c.show(show_ports=True)
> ```
> 
> it does not fillet some polyons, see bends on the right

That's because fillet is only applied to corners where both edges are long enough for the required radius. In the case of the bend, the segments the form the curve are very small (to shape the bend curvature) so we can't fillet them with the simple algorithm that's implemented. You'd need a more general operation (like the maximal area covered by a moving circle inside the shape), which we definitely don't have.
